### PR TITLE
Method overloads with default parameter values should not overlap

### LINKED ISSUE #653: Method overloads with default parameter values should not overlap
Replaces http://jira.sonarsource.com/browse/LINTVS-180

```
void Print(string[] messages) {...}  // Noncompliant
void Print(string[] messages, string delimiter = "n") {...}
```

When you call `Print(new []\{"abc"\});` which method will be called? The first one, but it is not obvious. So we should avoid these overloads with default parameter values. 

Note there is this rule: https://msdn.microsoft.com/en-us/library/ms182135.aspx, CA1026, saying that optional parameters should be avoided altogether because of CLS (non-)compliance. But I wouldn't go this far as optional parameters are really useful if you use named parameters in the calls: http://codeblog.jonskeet.uk/2015/06/03/backwards-compatibility-is-still-hard/

R# ID: *MethodOverloadWithOptionalParameter*

This rule could be covered by the http://jira.sonarsource.com/browse/RSPEC-3220, which covers overloads with `params` argument.

Implements https://jira.sonarsource.com/browse/RSPEC-3427

### PR DESCRIPTION
**Pull request by tvsonar**
Fixes #653

### CODE REVIEW

File: src/SonarLint.CSharp/Rules/MethodOverloadOptionalParameter.cs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+                            continue;
+                        }
+
+                        var defaultCanBeUsed =
+                            IsMoreParameterAvailableInConflicting(hidingInfo) ||
+                            !MethodsUsingSameParameterNames(hidingInfo);
+
+                        var isOtherFile = syntax.SyntaxTree.FilePath != hidingMethodSyntax.SyntaxTree.FilePath;
+
+                        c.ReportDiagnostic(Diagnostic.Create(Rule, syntax.GetLocation(),
```
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/MethodOverloadOptionalParameter.cs]: Should only report in non generated files

File: src/SonarLint.CSharp/Rules/MethodOverloadOptionalParameter.cs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                {
+                    var methodSymbol = c.Symbol as IMethodSymbol;
+                    if (methodSymbol == null ||
+                        PublicMethodWithMultidimensionalArrayBase.IsMethodOverrideOrInterfaceImplementation(methodSymbol) ||
+                        !methodSymbol.Parameters.Any(p => p.IsOptional))
+                    {
+                        return;
+                    }
+
+                    var methods = methodSymbol.ContainingType.GetMembers()
```
Comment (dbolkensteyn) [inline src/SonarLint.CSharp/Rules/MethodOverloadOptionalParameter.cs]: why not `GetMembers(methodSymbol.Name)`

File: src/SonarLint.CSharp/Rules/MethodOverloadOptionalParameter.cs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                        .OfType<IMethodSymbol>();
+
+                    var hidingInfos = new List<ParameterHidingMethodInfo>();
+
+                    var matchingNamedMethods = methods
+                        .Where(m => m.Name == methodSymbol.Name)
```
Comment (dbolkensteyn) [inline src/SonarLint.CSharp/Rules/MethodOverloadOptionalParameter.cs]: then this one could be dropped

File: src/SonarLint.CSharp/Rules/MethodOverloadOptionalParameter.cs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        {
+            return hidingInfo.HiddenMethod.Parameters.IndexOf(hidingInfo.ParameterToReportOn) <
+                hidingInfo.HiddenMethod.Parameters.Count() - 1;
+        }
+
+        private static bool IsMethodHidingOriginal(IMethodSymbol candidateHidingMethod, IMethodSymbol method)
+        {
+            for (int i = 0; i < candidateHidingMethod.Parameters.Length; i++)
+            {
+                if (!candidateHidingMethod.Parameters[i].Type.Equals(method.Parameters[i].Type))
```
Comment (dbolkensteyn) [inline src/SonarLint.CSharp/Rules/MethodOverloadOptionalParameter.cs]: iterating over parallel collections .. :)

**Comment by dbolkensteyn** [comment]
LGTM, tested

**Pull request merged**